common/src/main/java/org/apache/drill/common/CatastrophicFailure.java

```diff
@@ -46,6 +46,7 @@ public static void exit(Throwable e, String message, int code) {
       Thread.sleep(1000);
     } catch (InterruptedException e2) {
     }
-    System.exit(code);
+
+    Runtime.getRuntime().halt(code);
```

Comment:
I am not 100% convinced that using `halt()` it is the right approach. Note that by default, if OOM Error is thrown installed hooks are invoked by a JVM:
```
public class OOMTest {
  public static void main(String[] args) {
    Runtime.getRuntime().addShutdownHook(new Thread() {
      @Override
      public void run() {
        System.out.println("Hook");
      }
    });
    Object[] a = new Object[Integer.MAX_VALUE];
  }
}
```

Reply:
Yes. That is because the OOM is not caught. Drill catches OOMs and calls CatastrophicFailure.exit. If you catch the OOM and then hault the hooks don't get run. I have verified this behavior with this

```
  public static void main(String[] args) {
    Runtime.getRuntime().addShutdownHook(new Thread() {
      @Override
      public void run() {
        System.out.println("Hook");
      }
    });
    
    try {
      Object[] a = new Object[Integer.MAX_VALUE];
    } catch (OutOfMemoryError e) {
      Runtime.getRuntime().halt(0);
    }
  }
```

Reply:
My point is that default JVM behavior is that OOM does not bypass shutdown hook and other libraries may install shutdown hook that they expect will be called on JVM exit.

From Java documentation "This (halt) method should be used with extreme caution.", so I would prefer a solution that bypasses Drill close sequence, but still gives other libraries chance to invoke shutdown hooks.

Reply:
I think this boils down to what you consider a Heap OOM. In my mind if we experience a heap OOM we cannot continue executing any code safely, this includes any shutdown sequences. The reason why it is not safe is because the JVM can get stuck in gc during a memory allocation. So if we run any shutdown code we can hang, and that was the issue we were trying to fix in the first place. Additionally I'm not totally convinced we would want to bypass Drill's close sequence but run the Shutdown hooks for the libraries that Drill uses. That could lead to a dangerous situation where Drill continues running when some underlying resource are defunct and we can get even more errors and hangs.

At the end of the day though it's just your word against mine. So let's do an experiment. I'll go your route and add a flag to bypass only Drill's ShutdownThread in the event of a CatastrophicFailure. If the issue of hanging shutdowns never occurs again for users, great. If it continues to occur for users after this change, then we should agree to make the change to forcibly terminate the JVM with halt.

Reply:
Heap OOM means that JVM tried to allocate a new instance of some class and it failed for whatever reason (not enough heap available, GC taking too long and etc). It does not mean that JVM is fully unstable and will fail during GC. It means that any attempt to allocate another class may (but is not required as in my example if you handle OOM and instantiate a small object) fail as well. JVM always has reserved memory to handle OOM.

Your plan to try first a flag to bypass only Drill close sequence in the shutdown hook works for me.

Reply:
Also there is one more issue with both approaches I thought of. Currently, CatastrophicFailure.exit is only called when an OOM is explicitly caught in the Foreman and FragmentExecutor. However, these are not all the places where catastrophic failures can happen. If an OOM happens anywhere else we will continue to do the wrong thing and try to run the normal shutdown sequence. We should use https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#setUncaughtExceptionHandler-java.lang.Thread.UncaughtExceptionHandler- to set the default exception handler on all threads to properly call CatastrophicFailure.exit in the event of a heap OOM.

So I will implement this as well.